day_5_supply_stacks/stack_parser.rb

```diff
@@ -0,0 +1,78 @@
+require 'minitest/autorun'
+
+class StackParser
+  def initialize(stack_input)
+    @stack_input = stack_input
+    @stack_rows = stack_rows
+  end
+
+  def stack_columns
+    raw_stack_columns = transpose_stack_rows
+    remove_empty_crates_from(raw_stack_columns)
+  end
+
+  private
+
+  attr_reader :stack_input, :stack_row
+
+  def transpose_stack_rows
+    rows = stack_rows.size
+    columns = stack_rows.max_by(&:size).size
+
+    raw_stack_columns = Array.new(columns) { Array.new(rows) }
+
+    stack_rows.each_with_index do |row, i|
+      row.each_with_index do |crate, j|
+        raw_stack_columns[j][i] = crate
+      end
+    end
+
+    raw_stack_columns
+  end
+
+  def remove_empty_crates_from(raw_stack_columns)
+    raw_stack_columns.map do |column|
+      column.delete_if { |crate| crate.nil? || crate.empty? }
+    end
+  end
+
+  def stack_rows
+    stack_input.map do |row|
+      row.each_char.each_slice(4).map do |crate|
+        crate.join('')
+             .gsub("[", "")
+             .gsub("]","")
+             .strip
+      end
+    end
+  end
+end
+
+class TestStackParser < MiniTest::Test
+  def test_intial_stack_setup
+    stack_input2 =[
+      "[N]     [Q]         [N]\n",
+      "[R]     [F] [Q]     [G] [M]\n",
+      "[J]     [Z] [T]     [R] [H] [J]\n",
+      "[T] [H] [G] [R]     [B] [N] [T]\n",
+      "[Z] [J] [J] [G] [F] [Z] [S] [M]\n",
+      "[B] [N] [N] [N] [Q] [W] [L] [Q] [S]\n",
+      "[D] [S] [R] [V] [T] [C] [C] [N] [G]\n",
+      "[F] [R] [C] [F] [L] [Q] [F] [D] [P]\n",
+    ]
+
+    # Frist element is the top crate in the stack
+    columns_in_stack2 = [
+      ["N", "R", "J", "T", "Z", "B", "D", "F"],
+      ["H", "J", "N", "S", "R"],
+      ["Q", "F", "Z", "G", "J", "N", "R", "C"],
+      ["Q", "T", "R", "G", "N", "V", "F"],
+      ["F", "Q", "T", "L"],
+      ["N", "G", "R", "B", "Z", "W", "C", "Q"],
+      ["M", "H", "N", "S", "L", "C", "F"],
+      ["J", "T", "M", "Q", "N", "D"],
+      ["S", "G", "P"]]
```

Comment:
So hey, heredoc right! Jesus himself taught me! So why the fuck not!

```ruby
columns_in_stack2_alt = <<MAYBE_DONT_DO_THIS.split("\n").map{|row| row.split("")}
NRJTZBDF
HJNSR
QFZGJNRC
QTRGNVF
FQTL
NGRBZWCQ
MHNSLCF
JTMQND
MAYBE_DONT_DO_THIS
```

Hmm, not sure I like this - yeah it's a LOT less typing, but for the expected value of a test, it's a good idea to type out the literal expectation. 

Although for throwaway code like this, it would be OK.

But for prod code, esp. if the code is read by other people, I'd go with your version - it's immediately clear wtf is going on. The Jesus version requires some extra thinking, and if we can save that (the 'extra' part, that is, not thinking in general ;-) it's always a win!

Reply:
Interesting, but at the same time, confusing in this case. Tests (at least for me) about clarity: "Oh, so this is what we have, and we need to end up with." 

Later on, I opted for simply reading the test input from a file (the same way I do it with the real input). It is a lot cleaner code, but it doesn't reveal what are you working with. 

`yeah, it's a LOT less typing` -> actually, it is not. In this case, the input you see in the tests is a simple cut-and-paste from the input file of the riddle. I am way too lazy to type that much ;).